zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSetting.java

```diff
@@ -477,10 +478,22 @@ void closeInterpreters(String user, String noteId) {
 
   public void close() {
     LOGGER.info("Close InterpreterSetting: " + name);
-    for (ManagedInterpreterGroup intpGroup : interpreterGroups.values()) {
-      intpGroup.close();
-    }
+    List<Thread> closeThreads = interpreterGroups.values().stream()
+            .map(g -> new Thread(g::close, name + "-close"))
+            .peek(t -> t.setUncaughtExceptionHandler((th, e) ->
+                    LOGGER.error("InterpreterSetting close error", e)))
+            .peek(Thread::start)
+            .collect(Collectors.toList());
     interpreterGroups.clear();
+    for (Thread t : closeThreads) {
+      try {
+        t.join();
+      } catch (InterruptedException e) {
+        LOGGER.error("Can't wait InterpreterSetting close threads", e);
+        Thread.currentThread().interrupt();
+        break;
```

Comment:
instead of breaking, should it try to wait/close the other interpreter as best as it can?

Reply:
In the current implementation the state of `InterpreterSetting` will be eventually consistent, because:

- `interpreterGroups` are cleaned up before joining (No one will see the "partially closed object" in the list);
- `close-threads` will sooner or later finish their work.

And I don't see how it can harm.

Taking [these](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#2.1) [recommendations](https://rules.sonarsource.com/java/tag/multi-threading/RSPEC-2142) into account, in case of interruption we must:

- exit current thread asap;
- set the value of the `isInterrupted` flag back if we cannot stop current thread or rethrow `InterruptedException`.

To meet these requirements, we can `join()` all the threads and then set back the value of the flag (in case of `InterruptedException`). But, this is going to be a bit tricky. 
How important is it to wait for the completion of all the `close-threads`, in case of a forced interruption of the thread?